rust/hook-janitor/src/webhooks.rs

```diff
@@ -139,6 +134,126 @@ impl From<FailedRow> for AppMetric {
     }
 }
 
+#[derive(sqlx::FromRow, Debug)]
+struct HoghookCompletedRow {
+    // App Metrics truncates/aggregates rows on the hour, so we take advantage of that to GROUP BY
+    // and aggregate to select fewer rows.
+    hour: DateTime<Utc>,
+    // A note about the `try_from`s: Postgres returns all of those types as `bigint` (i64), but
+    // we know their true sizes, and so we can convert them to the correct types here. If this
+    // ever fails then something has gone wrong.
+    #[sqlx(try_from = "i64")]
+    team_id: u32,
+    app_source_id: String,
+    #[sqlx(try_from = "i64")]
+    count: u32,
+}
+
+impl From<HoghookCompletedRow> for AppMetric2 {
+    fn from(row: HoghookCompletedRow) -> Self {
+        AppMetric2 {
+            team_id: row.team_id,
+            timestamp: row.hour,
+            app_source: app_metrics2::Source::Hoghooks,
+            app_source_id: row.app_source_id,
+            instance_id: None,
+            metric_kind: app_metrics2::Kind::Success,
+            metric_name: "Fetch".to_owned(),
+            count: row.count,
+        }
+    }
+}
+
+#[derive(sqlx::FromRow, Debug)]
+struct HoghookFailedRow {
+    // App Metrics truncates/aggregates rows on the hour, so we take advantage of that to GROUP BY
+    // and aggregate to select fewer rows.
+    hour: DateTime<Utc>,
+    // A note about the `try_from`s: Postgres returns all of those types as `bigint` (i64), but
+    // we know their true sizes, and so we can convert them to the correct types here. If this
+    // ever fails then something has gone wrong.
+    #[sqlx(try_from = "i64")]
+    team_id: u32,
+    app_source_id: String,
+    #[sqlx(json)]
+    last_error: WebhookJobError,
+    #[sqlx(try_from = "i64")]
+    count: u32,
+}
+
+impl From<HoghookFailedRow> for AppMetric2 {
+    fn from(row: HoghookFailedRow) -> Self {
+        AppMetric2 {
+            team_id: row.team_id,
+            timestamp: row.hour,
+            app_source: app_metrics2::Source::Hoghooks,
+            app_source_id: row.app_source_id,
+            instance_id: None,
+            metric_kind: app_metrics2::Kind::Failure,
+            metric_name: row.last_error.r#type.to_string(),
+            count: row.count,
+        }
+    }
+}
+
+#[derive(sqlx::FromRow, Debug)]
+struct QueueDepth {
+    oldest_scheduled_at_untried: DateTime<Utc>,
+    count_untried: i64,
+    oldest_scheduled_at_retries: DateTime<Utc>,
+    count_retries: i64,
+}
+
+async fn send_metrics_to_kafka<T>(
+    kafka_producer: &FutureProducer<KafkaContext>,
+    topic: &str,
+    metrics: impl IntoIterator<Item = T>,
+) -> Result<()>
+where
+    T: Serialize,
```

Comment:
I do think this function is about 1 step from becoming a higher level utility `send_serializable_iterator_to_kafka`, I was just feeling lazy about extracting it since we don't have another use (*yet*).

With that in mind, would we still want a tighter trait bound?

Reply:
I think leaving it as is (with maybe a TODO to make it even more abstract) is a good idea.